File: src/retrieval/service.py
import asyncio
from dataclasses import dataclass, field
from typing import AsyncIterator
from sqlalchemy.ext.asyncio import AsyncSession
from src.retrieval.semantic import SemanticSearch
from src.retrieval.bm25 import BM25Search
from src.retrieval.rrf import rrf_merge
from src.retrieval.expand import expand_to_l2
from src.retrieval.reranker import Reranker
from src.retrieval.context import ContextBuilder
from src.llm.protocol import LLMProvider
from src.retrieval.memory_search import MemorySearch, MemoryHit
from src.profiling import StepTimer


@dataclass
class QueryResult:
    answer: str
    sources: list[dict] = field(default_factory=list)
    input_tokens: int = 0
    output_tokens: int = 0


class RetrievalService:
    def __init__(
        self,
        semantic_search: SemanticSearch,
        bm25_search: BM25Search,
        reranker: Reranker,
        context_builder: ContextBuilder,
        llm_provider: LLMProvider,
        rrf_k: int = 60,
        reranker_top_n: int = 5,
        memory_search: "MemorySearch | None" = None,
    ):
        self.semantic_search = semantic_search
        self.bm25_search = bm25_search
        self.reranker = reranker
        self.context_builder = context_builder
        self.llm_provider = llm_provider
        self.rrf_k = rrf_k
        self.reranker_top_n = reranker_top_n
        self.memory_search = memory_search

    async def query(
        self,
        session: AsyncSession,
        query: str,
        embed_fn,
        memory_search: "MemorySearch | None" = None,
    ) -> QueryResult:
        t = StepTimer("query")

        with t.step("embed"):
            query_vector = await embed_fn(query)

        with t.step("semantic"):
            semantic_hits = await self.semantic_search.search(session, query_vector)
        with t.step("bm25"):
            bm25_hits = await self.bm25_search.search(session, query)
        with t.step("expand"):
            merged = rrf_merge(semantic_hits, bm25_hits, k=self.rrf_k)
            l2_chunks = await expand_to_l2(session, merged)
        with t.step("rerank"):
            reranked = await self.reranker.rerank(query, l2_chunks, top_n=self.reranker_top_n)

        effective_memory_search = memory_search or self.memory_search
        mem_hits = []
        if effective_memory_search:
            with t.step("memory"):
                mem_hits = await effective_memory_search.search(session, query_vector)

        ctx = self.context_builder.build(query, reranked, memory_hits=mem_hits)

        with t.step("llm"):
            llm_response = await self.llm_provider.complete(ctx.prompt)

        t.log()
        return QueryResult(
            answer=llm_response.answer,
            sources=ctx.sources,
            input_tokens=llm_response.input_tokens,
            output_tokens=llm_response.output_tokens,
        )

    async def search_chunks(
        self,
        session: AsyncSession,
        query: str,
        embed_fn,
        top_k: int = 5,
    ) -> list[dict]:
        from pathlib import Path
        query_vector = await embed_fn(query)
        semantic_hits = await self.semantic_search.search(session, query_vector)
        bm25_hits = await self.bm25_search.search(session, query)
        merged = rrf_merge(semantic_hits, bm25_hits, k=self.rrf_k)
        l2_chunks = await expand_to_l2(session, merged)
        reranked = await self.reranker.rerank(query, l2_chunks, top_n=top_k)
        return [
            {
                "text": c.content,
                "doc_id": str(c.doc_id),
                "title": c.doc_title,
                "filename": Path(c.doc_source).name.split("-", 5)[-1] if c.doc_source else None,
                "section": c.section_heading,
                "page": c.page_number,
            }
            for c in reranked
        ]

    async def query_stream(
        self,
        session: AsyncSession,
        query: str,
        embed_fn,
        memory_search: "MemorySearch | None" = None,
    ) -> AsyncIterator[dict]:
        t = StepTimer("stream")

        with t.step("embed"):
            query_vector = await embed_fn(query)
        with t.step("semantic"):
            semantic_hits = await self.semantic_search.search(session, query_vector)
        with t.step("bm25"):
            bm25_hits = await self.bm25_search.search(session, query)
        with t.step("expand"):
            merged = rrf_merge(semantic_hits, bm25_hits, k=self.rrf_k)
            l2_chunks = await expand_to_l2(session, merged)
        with t.step("rerank"):
            reranked = await self.reranker.rerank(query, l2_chunks, top_n=self.reranker_top_n)

        effective_memory_search = memory_search or self.memory_search
        mem_hits = []
        if effective_memory_search:
            with t.step("memory"):
                mem_hits = await effective_memory_search.search(session, query_vector)

        ctx = self.context_builder.build(query, reranked, memory_hits=mem_hits)
        t.log()  # log before streaming starts

        async for token in self.llm_provider.complete_stream(ctx.prompt):
            yield {"type": "token", "data": token}

        yield {"type": "sources", "data": ctx.sources}
        yield {"type": "done"}

File: rust/src/search/service.rs
use std::sync::Arc;

use chrono::Local;
use rusqlite::Connection;

use crate::db::repositories::chunks::ChunkRepository;
use crate::ingestion::embeddings::EmbeddingClient;
use crate::ingestion::language::LanguageDetector;
use crate::llm::LlmProvider;
use crate::search::context::ContextBuilder;
use crate::search::expand::expand_to_l2;
use crate::search::memory_search::MemorySearch;
use crate::search::reranker::Reranker;
use crate::search::rrf::{rrf_merge, SearchHit};
use crate::search::tantivy_fts::{FtsHit, TantivyStore};
use crate::search::vectors::{VectorHit, VectorStore};

/// The result of a retrieval + generation query.
#[derive(Debug)]
pub struct QueryResult {
    pub answer: String,
    pub sources: Vec<serde_json::Value>,
    pub input_tokens: u32,
    pub output_tokens: u32,
}

/// Orchestrates the full retrieval pipeline:
/// embed → vector search → FTS → RRF merge → L2 expand → rerank → memory → LLM.
pub struct RetrievalService {
    pub embed: Arc<EmbeddingClient>,
    pub tantivy: Arc<TantivyStore>,
    pub vectors: Arc<VectorStore>,
    pub reranker: Arc<Reranker>,
    pub llm: Arc<dyn LlmProvider>,
    pub context_builder: ContextBuilder,
    pub memory_search: MemorySearch,
    pub lang: LanguageDetector,
    pub semantic_top_k: usize,
    pub bm25_top_k: usize,
    pub rrf_k: usize,
    pub reranker_top_n: usize,
}

impl RetrievalService {
    /// Execute the full retrieval + generation pipeline.
    ///
    /// This method is **synchronous** and is designed to be called from within
    /// `tokio::task::spawn_blocking`. The async LLM call is bridged via
    /// `tokio::runtime::Handle::current().block_on(...)`.
    pub fn query(
        &self,
        conn: &Connection,
        query: &str,
        memory_category: Option<&str>,
    ) -> anyhow::Result<QueryResult> {
        // Step 1: Embed the query
        let query_vector = self.embed.embed_query(query)?;

        // Step 2: Vector (semantic) search
        let raw_vector_hits = self.vectors.search_chunks(conn, &query_vector, self.semantic_top_k)?;

        // Step 3: Load chunk metadata for each vector hit
        let semantic_hits: Vec<SearchHit> = raw_vector_hits
            .iter()
            .filter_map(|h| load_chunk_for_vector_hit(conn, h))
            .collect();

        // Step 4: BM25 / FTS search
        let lang = self.lang.detect(query);
        let raw_fts_hits = self.tantivy.search(query, &lang, self.bm25_top_k)?;

        // Step 5: Load chunk metadata for each FTS hit
        let bm25_hits: Vec<SearchHit> = raw_fts_hits
            .iter()
            .filter_map(|h| load_chunk_for_fts_hit(conn, h))
            .collect();

        // Step 6: RRF merge
        let merged = rrf_merge(&semantic_hits, &bm25_hits, self.rrf_k, 20);

        // Step 7: Expand to L2 (leaf → parent chunks)
        let l2_chunks = expand_to_l2(conn, &merged)?;

        // Step 8: Rerank
        let texts: Vec<&str> = l2_chunks.iter().map(|c| c.content.as_str()).collect();
        let reranked_chunks = if !texts.is_empty() {
            let rerank_results = self.reranker.rerank(query, &texts, self.reranker_top_n)?;
            rerank_results
                .iter()
                .filter_map(|r| l2_chunks.get(r.original_index))
                .cloned()
                .collect::<Vec<_>>()
        } else {
            l2_chunks
        };

        // Step 9: Memory search
        let mem_hits =
            self.memory_search.search(conn, &self.vectors, &query_vector, memory_category)?;

        // Step 10: Build context
        let today = Local::now().format("%Y-%m-%d").to_string();
        let ctx = self.context_builder.build(query, &reranked_chunks, &mem_hits, &today);

        // Step 11: Call LLM (async, bridged via Handle::current().block_on)
        let llm = Arc::clone(&self.llm);
        let prompt = ctx.prompt.clone();
        let llm_response = tokio::runtime::Handle::current()
            .block_on(async move { llm.complete(&prompt).await })?;

        Ok(QueryResult {
            answer: llm_response.answer,
            sources: ctx.sources,
            input_tokens: llm_response.input_tokens,
            output_tokens: llm_response.output_tokens,
        })
    }
}

/// Load a [`SearchHit`] from the DB for a vector search result.
fn load_chunk_for_vector_hit(conn: &Connection, hit: &VectorHit) -> Option<SearchHit> {
    let repo = ChunkRepository::new(conn);
    let chunks = repo.get_by_ids(std::slice::from_ref(&hit.id)).ok()?;
    let chunk = chunks.into_iter().next()?;

    Some(SearchHit {
        chunk_id: chunk.id,
        content: chunk.content,
        parent_chunk_id: chunk.parent_chunk_id,
        doc_id: chunk.doc_id,
        score: 1.0 - hit.distance, // convert L2 distance to rough similarity
        section_heading: chunk.section_heading,
        page_number: chunk.page_number.map(|p| p as u32),
    })
}

/// Load a [`SearchHit`] from the DB for an FTS search result.
fn load_chunk_for_fts_hit(conn: &Connection, hit: &FtsHit) -> Option<SearchHit> {
    let repo = ChunkRepository::new(conn);
    let chunks = repo.get_by_ids(std::slice::from_ref(&hit.chunk_id)).ok()?;
    let chunk = chunks.into_iter().next()?;

    Some(SearchHit {
        chunk_id: chunk.id,
        content: chunk.content,
        parent_chunk_id: chunk.parent_chunk_id,
        doc_id: chunk.doc_id,
        score: hit.score,
        section_heading: chunk.section_heading,
        page_number: chunk.page_number.map(|p| p as u32),
    })
}
